numba/np/arraymath.py

```diff
@@ -669,6 +670,111 @@ def _trivial_isnan(x):
         return _trivial_isnan
 
 
+@register_jitable
+def _impl_iscomplex_isreal_arr(x, fn):
+    ax = np.asarray(x) # NumPy uses asanyarray here!
+    return fn(ax.imag, 0)
+
+
+@overload(np.iscomplex)
+def np_iscomplex(x):
+    if type_can_asarray(x):
+        return lambda x: _impl_iscomplex_isreal_arr(x, operator.ne)
+    return None
+
+
+@overload(np.isreal)
+def np_isreal(x):
+    if type_can_asarray(x):
+        return lambda x: _impl_iscomplex_isreal_arr(x, operator.eq)
+    return None
+
+
+@overload(np.iscomplexobj)
+def iscomplexobj(x):
+    # Implementation based on NumPy
+    # https://github.com/numpy/numpy/blob/d9b1e32cb8ef90d6b4a47853241db2a28146a57d/numpy/lib/type_check.py#L282-L320
+    if isinstance(x, types.Optional):
+        msg = 'Expected argument {} but got Optional[{}, None]'
+        raise TypingError(msg.format(x.key, x.key))
+    dt = determine_dtype(x)
+    iscmplx = np.issubdtype(dt, np.complexfloating)
+    return lambda x: False if x is None else iscmplx
```

Comment:
Should this be simply:
```suggestion
    return lambda x: iscmplx
```